analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Constraints/NumberConstraint.cs

```diff
@@ -85,18 +66,24 @@ value switch
 
     public static NumberConstraint From(BigInteger? min, BigInteger? max)
     {
-        if (min.HasValue && max.HasValue && min.Value == max.Value)
+        if (!min.HasValue && !max.HasValue)
         {
-            return From(min.Value);
+            return null;
         }
-        else if (min.HasValue || max.HasValue)
+        if (min.HasValue && max.HasValue && min.Value > max.Value)
         {
-            return new(min, max);
+            var tmp = min;
+            min = max;
+            max = tmp;
         }
-        else
+        if (cache.Count > CacheLimit)
         {
-            return null;
+            ResetCache();
         }
+        var key = new CacheKey(min, max);
+        return cache.TryGetValue(key, out var constraint)
+            ? constraint
+            : cache.GetOrAdd(key, new NumberConstraint(min, max));
```

Comment:
To avoid allocations of the lambda that needs to be passed. These tends to create a lot of objects on the critical paths.

Yes, this can go out of sync a little, by creating a few useless instances. Practically, it's not a problem because it doesn't matter who is the winner instance.

Reply:
The lambda is de-sugared into a lazy-initialized singleton (e.g. [here](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQAYACdbAOgGEB7AGzoFMBjAFyhpgGdrPmBXAE6DGMVgG4CBdAGYymElRIBvAiXVk56ACwkA6lFYALGv1YBlVhHbMAFLRgDho1gBEobDjAiCAngB5yIjgybCIAPhJmCGYjRgBKFTUNFOjYxgoAcUZWAHlBAEEEBFsAIgBrRl8ANQg6UpDK3xIAXkjSgDc60vjJfBSAX2T1YdkyXQNjS2sPez4hETF3T04fAKCQoMi0uMTVfpSNHYzsvMLisqba+s3sADYSJtb2rvre4ZIBkgIBoA=) ).

The only price paid after the first call is checking whether the `static readonly` field is initialized or not, which is very little.

Even assuming closure of `min` and `max`, that would boil down to setting a mutable field on the singleton:
```cs
    [System.Runtime.CompilerServices.NullableContext(1)]
    public void WithoutStatic(ConcurrentDictionary<string, string> cache)
    {
        <>c__DisplayClass0_0 <>c__DisplayClass0_ = new <>c__DisplayClass0_0();
        <>c__DisplayClass0_.min = 2;
        cache.GetOrAdd("keyVal", new Func<string, string>(<>c__DisplayClass0_.<WithoutStatic>b__0));
    }
```

And we can avoid this, since `GetOrAdd` conveniently passes the key as input parameter of the lambda, allowing for:
```cs
return cache.GetOrAdd(new CacheKey(min, max), static key => new NumberConstraint(key.Min, key.Max));
```

And that is even faster.

Reply:
It still creates `new Func<string, string>` that needs to be garbage collected. I think Martin was measuring the count of objects when he was doing the SymbolicValue caching.